src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexBoyerMoore.cs

```diff
@@ -229,7 +239,17 @@ public bool IsMatch(string text, int index, int beglimit, int endlimit)
 
             if (CaseInsensitive)
             {
-                return string.Compare(Pattern, 0, text, index, Pattern.Length, ignoreCase: true, _culture) == 0;
+                TextInfo textinfo = _culture.TextInfo;
+
+                for (int i = 0; i < Pattern.Length; i++)
+                {
+                    if (Pattern[i] != textinfo.ToLower(text[index + i]))
```

Comment:
>textinfo.ToLower(text[index + i]) [](start = 38, length = 33)

wouldn't be faster if we convert the whole text (with the length of pattern) to lowercase as one shot? calling ToLower with every character can be very expensive if the character is not ASCII. 

The other idea is to call CharUnicodeInfo.GetUnicodeCategory on the char before calling tolower so we avoid teh overhead of characters that really don't need to be lower cased.

Reply:
> wouldn't be faster if we convert the whole text (with the length of pattern) to lowercase as one shot?

How do we do that in a way that produces the same result and doesn't allocate?

> The other idea is to call CharUnicodeInfo.GetUnicodeCategory on the char before calling tolower so we avoid the overhead of characters that really don't need to be lower cased.

That will presumably add overhead for characters that do need to be lower-cased, correct?  Or else why wouldn't we just do that as part of the ToLower implementation?

Reply:
> How do we do that in a way that produces the same result and doesn't allocate?

I expect the pattern length most of the time is small length (less than 200 characters or so). we can stackalloc and Span to do that I guess. 

> That will presumably add overhead for characters that do need to be lower-cased, correct? Or else why wouldn't we just do that as part of the ToLower implementation?

It depends if all characters we are using are uppercased. I am thinking the number of characters in Unicode that need casing compared to the whole set of Unicode characters is a really small percentage (and we'll be interested in the uppercased letters only too). 
Also CharUnicodeInfo.GetUnicodeCategory would cost a very small percentage compared to the time in ToLower (in non-ASCII cases).

Reply:
> I expect the pattern length most of the time is small length (less than 200 characters or so). we can stackalloc and Span to do that I guess

And with what API?

Reply:
> And with what API? i.e. what API does culture-aware lower-casing on a whole span/string with behavior equivalent to doing it character-by-character?

Currently, the casing (including the culture-aware one) always done in one-to-one matching the number of characters in the string. It doesn't collapse or expand any character. We can add simple API TextInfo.ToLower which works with the Span and I believe these APIs will be very useful in many other cases. 
By the way, you don't have to block the PR for that. I am just pointing we can optimize more in this case.

Reply:
> We can add simple API TextInfo.ToLower which works with the Span and I believe these APIs will be very useful in many other cases.

If you think they'd be useful in other circumstances, I have no qualms with such methods being added.  Please open an issue for that, and if they're added, part of that work item should be updating this and any other possible call sites.

I was confirming that no such methods exist today, for span or string or any other input, correct?

Reply:
> I was confirming that no such methods exist today, for span or string or any other input, correct?

I am not aware of any although I recall this was requested before but I didn't find any issue for that.

Reply:
Great, thanks.